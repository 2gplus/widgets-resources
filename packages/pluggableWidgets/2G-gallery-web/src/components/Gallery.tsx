import { createElement, ReactElement, ReactNode } from "react";
import { InfiniteBody, Pagination } from "@mendix/piw-utils-internal/components/web";
import { ObjectItem } from "mendix";
import classNames from "classnames";
import { InfiniteBodyWidth } from "./InfiniteBodyWidth";

export interface GalleryProps<T extends ObjectItem> {
    className?: string;
    desktopItems: number;
    desktopItemsDynamic: boolean;
    emptyPlaceholderRenderer?: (renderWrapper: (children: ReactNode) => ReactElement) => ReactElement;
    emptyMessageTitle?: string;
    filters?: ReactNode;
    filtersTitle?: string;
    hasMoreItems: boolean;
    items: T[];
    itemRenderer: (
        renderWrapper: (children: ReactNode, className?: string, onClick?: () => void) => ReactElement,
        item: T
    ) => ReactNode;
    numberOfItems?: number;
    paging: boolean;
    verticalScroll: boolean;
    page: number;
    pageSize: number;
    paginationPosition?: "below" | "above";
    preview?: boolean;
    phoneItems: number;
    phoneItemsDynamic: boolean;
    setPage?: (computePage: (prevPage: number) => number) => void;
    tabletItems: number;
    tabletItemsDynamic: boolean;
    tabIndex?: number;
}

export function Gallery<T extends ObjectItem>(props: GalleryProps<T>): ReactElement {
    const pagination = props.paging ? (
        <div className="widget-gallery-pagination">
            <Pagination
                canNextPage={props.hasMoreItems}
                canPreviousPage={props.page !== 0}
                gotoPage={(page: number) => props.setPage && props.setPage(() => page)}
                nextPage={() => props.setPage && props.setPage(prev => prev + 1)}
                numberOfItems={props.numberOfItems}
                page={props.page}
                pageSize={props.pageSize}
                previousPage={() => props.setPage && props.setPage(prev => prev - 1)}
            />
        </div>
    ) : null;
    const infiniteChildren =
        props.items.length > 0 && props.itemRenderer
            ? props.items.map(item =>
                  props.itemRenderer((children, className, onClick) => {
                      return (
                          <div
                              key={`item_${item.id}`}
                              className={classNames("widget-gallery-item", className, {
                                  "widget-gallery-clickable": !!onClick
                              })}
                              onClick={onClick}
                              onKeyDown={
                                  onClick
                                      ? e => {
                                            if (e.key === "Enter" || e.key === " ") {
                                                e.preventDefault();
                                                onClick();
                                            }
                                        }
                                      : undefined
                              }
                              role={onClick ? "button" : "listitem"}
                              tabIndex={onClick ? 0 : undefined}
                          >
                              {children}
                          </div>
                      );
                  }, item)
              )
            : [];
    const getColumnWidth = (items: number, dynamic: boolean): number => {
        let retVal = items;
        if (dynamic && !props.hasMoreItems && infiniteChildren.length < items) {
            retVal = infiniteChildren.length;
        }
        return retVal;
    };
    const infiniteClassNames = (): string[] => {
        const retVal = ["widget-gallery-items"];
        if (!props.paging && props.verticalScroll) {
            retVal.push("horizontal-scroll");
        }
        retVal.push(`widget-auto-column-lg-${getColumnWidth(props.desktopItems, props.desktopItemsDynamic)}`);
        retVal.push(`widget-auto-column-md-${getColumnWidth(props.tabletItems, props.tabletItemsDynamic)}`);
        retVal.push(`widget-auto-column-sm-${getColumnWidth(props.phoneItems, props.phoneItemsDynamic)}`);
        return retVal;
    };
    return (
        <div className={classNames("widget-gallery", props.className)} data-focusindex={props.tabIndex || 0}>
            {props.paginationPosition === "above" && pagination}
            <div className="widget-gallery-filter" role="section" aria-label={props.filtersTitle}>
                {props.filters}
            </div>

            {infiniteChildren.length > 0 &&
                (props.verticalScroll ? (
                    <InfiniteBodyWidth
                        className={classNames(infiniteClassNames())}
                        hasMoreItems={props.hasMoreItems}
                        setPage={props.setPage}
                        isInfinite={!props.paging}
                        role="list"
                    >
                        {infiniteChildren}
                    </InfiniteBodyWidth>
                ) : (
                    <InfiniteBody
                        className={classNames(infiniteClassNames())}
                        hasMoreItems={props.hasMoreItems}
                        setPage={props.setPage}
                        isInfinite={!props.paging}
                        role="list"
                    >
                        {" "}
                        {infiniteChildren}
                    </InfiniteBody>
                ))}
            {(props.items.length === 0 || props.preview) &&
                props.emptyPlaceholderRenderer &&
                props.emptyPlaceholderRenderer(children => (
                    <div className="widget-gallery-empty" role="section" aria-label={props.emptyMessageTitle}>
                        <div className="empty-placeholder">{children}</div>
                    </div>
                ))}
            {props.paginationPosition === "below" && pagination}
        </div>
    );
}
