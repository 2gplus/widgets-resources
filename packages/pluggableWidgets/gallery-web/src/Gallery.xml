<?xml version="1.0" encoding="utf-8"?>
<widget id="com.mendix.widget.web.gallery.Gallery" pluginWidget="true" offlineCapable="true" xmlns="http://www.mendix.com/widget/1.0/" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.mendix.com/widget/1.0/ ../../../../node_modules/mendix/custom_widget.xsd">
    <name>Gallery</name>
    <description/>
    <studioProCategory>Data</studioProCategory>
    <studioCategory>Data</studioCategory>
    <icon>iVBORw0KGgoAAAANSUhEUgAAABgAAAAYCAYAAADgdz34AAAACXBIWXMAAAsTAAALEwEAmpwYAAAAAXNSR0IArs4c6QAAAARnQU1BAACxjwv8YQUAAAKLSURBVHgBvVUxbFJRFL0PSKWgCV2YTPhdYaEDc1mMlThIAoR0aNhZusAKnTRMusDamDBB8nEgqWk0dcfw1QSbOPiXDsKCESgR5HnvK4/ygf9tI3qS/9+7j/fuue/ce/kA/xiMXslkUhmPxylYAxhjeqVSOZa2g17kHH/IwfpwbCCQQGYm5w/UXvA0eleTdljlHicMPSfRTd1sTywWyy8GagMT2Jm9+UgdHErbCYNDPNy02rMKNqsfOXCPlW22Ng8hibzaj/2X25zxJgNmeWhlMJy/ulc+0MjPvNSGHAxh2HXyjRcUFzAbaXkGfPLuKhTbLr7DaB9dh3e9xw68hvMni8QGgrPoVheHPM33apc5OngSdV/Zaj+PDsPSXrUHlVgisMzBOuBYXHiIVcFk4lAWEfl0TgPZtvY5OD+oMDp/DfbOl13K4WTrvs66F75Ff4Ykd/fLO3b2q2kWjf3bZ7jzqSZGK4xGo+1araYv3eA0uqFJUtSXU0Klvo9L7/Obb5/lvF4vRFIp8Pv9QHOXyyXO6rounmq1Cu12+2s8Ht/BatIccEM43zylPxrIZDLg8y0pAYqizJ5sNktlSwm/OYFEoVAAjG7mjDAYDChqaLVaUK/XDfsNBHvqpWJoNKxzUYqInx9V1F8VshSLRdMAAoEAdDqd1QRWjWb7fiEajSTq9/szzQlut1vkgvKCxQLpdHo1gVWjUaXhaUEgJQqFQgaJGo2GtUTzYJzjTRza4vpfSSSilCgf0G2CuBYUhIyJRsvlcsKBmUQUQCKRMBLQZ246/vGrRiUYiUSWJOr1eqKKSqUSGHzCLYDap+gS+CgW27rYA0fYcM9vTTBHFJxMJgpGGZRr04g16l74n/gNNw874D7WvLUAAAAASUVORK5CYII=</icon>
    <properties>
        <propertyGroup caption="General">
            <propertyGroup caption="Advanced">
                <property key="advanced" type="boolean" defaultValue="false">
                    <caption>Advanced</caption>
                    <description />
                </property>
            </propertyGroup>
            <propertyGroup caption="General">
                <property key="datasource" type="datasource" isList="true">
                    <caption>Data source</caption>
                    <description/>
                </property>
                <property key="content" type="widgets" dataSource="datasource" required="false">
                    <caption>Content placeholder</caption>
                    <description/>
                </property>
            </propertyGroup>
            <propertyGroup caption="Columns">
                <property key="desktopItems" type="integer" defaultValue="1">
                    <caption>Desktop columns</caption>
                    <description />
                </property>
                <property key="tabletItems" type="integer" defaultValue="1">
                    <caption>Tablet columns</caption>
                    <description />
                </property>
                <property key="phoneItems" type="integer" defaultValue="1">
                    <caption>Phone columns</caption>
                    <description />
                </property>
            </propertyGroup>
            <propertyGroup caption="Items">
                <property key="pageSize" type="integer" defaultValue="20">
                    <caption>Page size</caption>
                    <description />
                </property>
                <property key="pagination" type="enumeration" defaultValue="buttons">
                    <caption>Pagination</caption>
                    <description/>
                    <enumerationValues>
                        <enumerationValue key="buttons">Paging buttons</enumerationValue>
                        <enumerationValue key="virtualScrolling">Virtual scrolling</enumerationValue>
                    </enumerationValues>
                </property>
                <property key="pagingPosition" type="enumeration" defaultValue="below">
                    <caption>Position of paging buttons</caption>
                    <description/>
                    <enumerationValues>
                        <enumerationValue key="below">Below grid</enumerationValue>
                        <enumerationValue key="above">Above grid</enumerationValue>
                    </enumerationValues>
                </property>
                <property key="showEmptyPlaceholder" type="enumeration" defaultValue="none">
                    <caption>Empty message</caption>
                    <description/>
                    <enumerationValues>
                        <enumerationValue key="none">None</enumerationValue>
                        <enumerationValue key="custom">Custom</enumerationValue>
                    </enumerationValues>
                </property>
                <property key="emptyPlaceholder" type="widgets" required="false">
                    <caption>Empty placeholder</caption>
                    <description/>
                </property>
                <property key="itemClass" type="expression" required="false" dataSource="datasource">
                    <caption>Dynamic item class</caption>
                    <description />
                    <returnType type="String"/>
                </property>
            </propertyGroup>
            <propertyGroup caption="Events">
                <property key="onClick" type="action" required="false" dataSource="datasource">
                    <caption>On click action</caption>
                    <description />
                </property>
            </propertyGroup>
        </propertyGroup>
        <propertyGroup caption="Filtering">
            <propertyGroup caption="Filtering">
                <property key="filterList" type="object" isList="true" required="false">
                    <caption>Filters</caption>
                    <description />
                    <properties>
                        <propertyGroup caption="General">
                            <property key="filter" type="attribute" dataSource="../datasource">
                                <caption>Filter attribute</caption>
                                <description />
                                <attributeTypes>
                                    <attributeType name="String"/>
                                    <attributeType name="AutoNumber"/>
                                    <attributeType name="Boolean"/>
                                    <attributeType name="DateTime"/>
                                    <attributeType name="Decimal"/>
                                    <attributeType name="Enum"/>
                                    <attributeType name="Integer"/>
                                    <attributeType name="Long"/>
                                </attributeTypes>
                            </property>
                        </propertyGroup>
                    </properties>
                </property>
                <property key="filtersPlaceholder" type="widgets" required="false">
                    <caption>Filters placeholder</caption>
                    <description/>
                </property>
            </propertyGroup>
        </propertyGroup>
        <propertyGroup caption="Sorting">
            <propertyGroup caption="Sorting">
                <property key="sortList" type="object" isList="true" required="false">
                    <caption>Sort attributes</caption>
                    <description />
                    <properties>
                        <propertyGroup caption="General">
                            <property key="attribute" type="attribute" dataSource="../datasource">
                                <caption>Attribute</caption>
                                <description />
                                <attributeTypes>
                                    <attributeType name="String"/>
                                    <attributeType name="AutoNumber"/>
                                    <attributeType name="Boolean"/>
                                    <attributeType name="DateTime"/>
                                    <attributeType name="Decimal"/>
                                    <attributeType name="Enum"/>
                                    <attributeType name="Integer"/>
                                    <attributeType name="Long"/>
                                </attributeTypes>
                            </property>
                            <property key="caption" type="textTemplate">
                                <caption>Caption</caption>
                                <description />
                            </property>
                        </propertyGroup>
                    </properties>
                </property>
            </propertyGroup>
        </propertyGroup>
        <propertyGroup caption="Accessibility">
            <propertyGroup caption="Accessibility">
                <property key="filterSectionTitle" type="textTemplate" required="false">
                    <caption>Filter section title</caption>
                    <description>Assistive technology will read this upon reaching a filtering or sorting section.</description>
                </property>
                <property key="emptyMessageTitle" type="textTemplate" required="false">
                    <caption>Empty message title</caption>
                    <description>Assistive technology will read this upon reaching an empty message section.</description>
                </property>
            </propertyGroup>
        </propertyGroup>
    </properties>
</widget>
