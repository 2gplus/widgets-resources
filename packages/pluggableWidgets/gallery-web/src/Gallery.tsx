import { createElement, ReactElement, useCallback, useEffect, useMemo, useRef, useState } from "react";
import { GalleryContainerProps } from "../typings/GalleryProps";
import { Gallery as GalleryComponent } from "./components/Gallery";
import "./ui/gallery-main.scss";
import { FilterType, useFilterContext, useMultipleFiltering } from "@mendix/piw-utils-internal";
import { FilterCondition } from "mendix/filters";
import { extractFilters } from "./utils/filters";
import { and } from "mendix/filters/builders";

export function Gallery(props: GalleryContainerProps): ReactElement {
    const viewStateFilters = useRef<FilterCondition | undefined>(undefined);
    const [filtered, setFiltered] = useState(false);
    const customFiltersState = useMultipleFiltering();
    const { FilterContext } = useFilterContext();

    useEffect(() => {
        if (props.datasource.filter && !filtered && !viewStateFilters.current) {
            viewStateFilters.current = props.datasource.filter;
        }
    }, [props.datasource, filtered]);

    const filterList = useMemo(
        () =>
            props.filterList
                .map(({ filter }) => ({ [filter.id]: filter }))
                .reduce((filters, current) => ({ ...filters, ...current }), {}),
        [props.filterList]
    );

    const initialFilters = useMemo(
        () =>
            props.filterList
                .map(filter => ({ [filter.filter.id]: extractFilters(filter.filter, viewStateFilters.current) }))
                .reduce((filters, current) => ({ ...filters, ...current }), {}),
        [props.filterList, viewStateFilters.current]
    );

    const filters = Object.keys(customFiltersState)
        .map((key: FilterType) => customFiltersState[key][0]?.getFilterCondition())
        .filter((filter): filter is FilterCondition => filter !== undefined);

    if (filters.length > 0) {
        props.datasource.setFilter(filters.length > 1 ? and(...filters) : filters[0]);
    } else if (filtered) {
        props.datasource.setFilter(undefined);
    } else {
        props.datasource.setFilter(viewStateFilters.current);
    }

    return (
        <GalleryComponent
            className={props.class}
            desktopItems={props.desktopItems}
            filters={useMemo(
                () => (
                    <FilterContext.Provider
                        value={{
                            filterDispatcher: prev => {
                                if (prev.filterType) {
                                    const [, filterDispatcher] = customFiltersState[prev.filterType];
                                    filterDispatcher(prev);
                                    setFiltered(true);
                                }
                                return prev;
                            },
                            multipleAttributes: filterList,
                            multipleInitialFilters: initialFilters
                        }}
                    >
                        {props.filtersPlaceholder}
                    </FilterContext.Provider>
                ),
                [FilterContext, customFiltersState, filterList, initialFilters, props.filtersPlaceholder]
            )}
            items={props.datasource.items ?? []}
            itemRenderer={useCallback(
                (renderWrapper, item) =>
                    renderWrapper(
                        props.content.get(item),
                        props.itemClass?.get(item)?.value,
                        props.onClick ? props.onClick?.get(item).execute : undefined
                    ),
                [props.content, props.itemClass, props.onClick]
            )}
            phoneItems={props.phoneItems}
            tabletItems={props.tabletItems}
            tabIndex={props.tabIndex}
        />
    );
}
