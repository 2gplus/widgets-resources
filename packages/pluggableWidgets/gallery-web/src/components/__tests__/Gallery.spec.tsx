import { createElement } from "react";
import { shallow } from "enzyme";
import { Gallery } from "../Gallery";

describe("Gallery", () => {
    it("renders correctly", () => {
        const gallery = shallow(
            <Gallery
                items={["item1"]}
                itemRenderer={(wrapper, item) => wrapper(item)}
                tabletItems={3}
                desktopItems={4}
                phoneItems={2}
            />
        );

        expect(gallery).toMatchSnapshot();
    });
});
