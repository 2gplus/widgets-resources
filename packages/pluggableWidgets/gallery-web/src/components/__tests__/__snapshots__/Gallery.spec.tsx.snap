// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Gallery DOM Structure renders correctly 1`] = `
<div
  class="widget-gallery"
  data-focusindex="0"
>
  <div
    class="widget-gallery-filter"
  />
  <div
    class="widget-gallery-items widget-gallery-lg-4 widget-gallery-md-3 widget-gallery-sm-2 infinite-loading"
  >
    <div
      class="widget-gallery-item"
    >
      11
    </div>
    <div
      class="widget-gallery-item"
    >
      22
    </div>
    <div
      class="widget-gallery-item"
    >
      33
    </div>
  </div>
</div>
`;

exports[`Gallery DOM Structure renders correctly with onclick event 1`] = `
<div
  class="widget-gallery"
  data-focusindex="0"
>
  <div
    class="widget-gallery-filter"
  />
  <div
    class="widget-gallery-items widget-gallery-lg-4 widget-gallery-md-3 widget-gallery-sm-2 infinite-loading"
  >
    <div
      class="widget-gallery-item widget-gallery-clickable"
      role="button"
      tabindex="0"
    >
      11
    </div>
    <div
      class="widget-gallery-item widget-gallery-clickable"
      role="button"
      tabindex="0"
    >
      22
    </div>
    <div
      class="widget-gallery-item widget-gallery-clickable"
      role="button"
      tabindex="0"
    >
      33
    </div>
  </div>
</div>
`;

exports[`Gallery with empty option renders correctly 1`] = `
<div
  class="widget-gallery"
  data-focusindex="0"
>
  <div
    class="widget-gallery-filter"
  />
  <div
    class="widget-gallery-empty"
  >
    <div
      class="empty-placeholder"
    >
      <span>
        No items found
      </span>
    </div>
  </div>
</div>
`;

exports[`Gallery with pagination renders correctly 1`] = `
<div
  class="widget-gallery"
  data-focusindex="0"
>
  <div
    class="widget-gallery-pagination"
  >
    <div
      class="pagination-bar"
    >
      <button
        class="btn pagination-button"
        disabled=""
      >
        <span
          class="glyphicon glyphicon-step-backward"
        />
      </button>
      <button
        class="btn pagination-button"
        disabled=""
      >
        <span
          class="glyphicon glyphicon-backward"
        />
      </button>
      <div
        class="paging-status"
      >
        1 to 10 of 20
      </div>
      <button
        class="btn pagination-button"
      >
        <span
          class="glyphicon glyphicon-forward"
        />
      </button>
      <button
        class="btn pagination-button"
      >
        <span
          class="glyphicon glyphicon-step-forward"
        />
      </button>
    </div>
  </div>
  <div
    class="widget-gallery-filter"
  />
  <div
    class="widget-gallery-items widget-gallery-lg-4 widget-gallery-md-3 widget-gallery-sm-2"
  >
    <div
      class="widget-gallery-item"
    >
      11
    </div>
    <div
      class="widget-gallery-item"
    >
      22
    </div>
    <div
      class="widget-gallery-item"
    >
      33
    </div>
  </div>
</div>
`;
